### ISSUE TITLE
[Feature request] sp-kill-whole-line

### ISSUE THREAD

**Issue by jabranham** [Fuco1/smartparens#674]
It would be nice to have the function `sp-kill-whole-line` that would kill the line that point is currently on but leave any necessary closing parens/braces/whatever. So that (with `|` representing point)

```
(something
    (something | else))
```
would become 

```
(something
   )
```

And then when `strict-mode` is enabled, it can be added to the remaps in `smartparens-strict-mode-map` for Emacs' default `kill-whole-line`.

**Comment by Fuco1** [comment]
Would the proposal in #169 work for you?  It is 3 years old but if more people are requesting this we might schedule it for some future release (most likely 1.11)

**Comment by Fuco1** [comment]
As a test... would this be equivalent to just

1. go to beginning of line
2. call `sp-kill-hybrid-sexp`

? If so we can implement it like so immediately.  Try to come up with some examples where it wouldn't do what you expect.

**Comment by jabranham** [comment]
I think your suggestion (beginning of line, then `sp-kill-hybdir-sexp`) would work. At least, I can't think of any examples where it wouldn't...

**Comment by Fuco1** [comment]
Should we than have two commands, or respect the variable setting?  To be honest I don't use this "workflow" much so I can't judge myself which would be more useful.

**Comment by Yevgnen** [comment]
I use these two commands

``` 
  (defun sp-change-inner ()
    "Smartparens version of `change-inner', kill content between the nearest pair immediately."
    (interactive)
    (sp-beginning-of-sexp)
    (call-interactively 'set-mark-command)
    (sp-end-of-sexp)
    (kill-region (region-beginning) (region-end)))

  (defun sp-change-outer ()
    "Smartparens version of `change-outer', kill content and the nearest pair immediately."
    (interactive)
    (sp-change-inner)
    (sp-backward-delete-char))
```

bound to `M-i` and `M-o` as an aggressive version of `change-inner` and `change-outer` to kill thing like that since `smartparens` can manage pairs more powerful that `change-inner` which uses `expand-region`. For @jabranham 's case, it's just a `M-o` press. 

I'm not familiar with lisp. But maybe a more subtle implement can be added to `smartparens` ?

**Comment by Fuco1** [comment]
@Yevgnen 
You should never use mark for computation in code.  You should rather save the `(point)` in a variable.

In smartparens, you can call `sp-get-enclosing-sexp` to get the data, and use `sp-get` macro to extract the result.

The entire function then results to

```
(sp-get (sp-get-enclosing-sexp)
  (kill-region :beg-in :end-in))  ; or :beg :end to kill delimiters too
```

Finally, `sp-kill-sexp` with `C-0` is exactly `sp-change-inner` and with `C-u` it is `sp-change-outer` :)


The difference with hybrid kill is that in a buffer like

```
(foo
 |bar asdad
 baz)
```

only the second line will be killed instead of the entire sexp (or its body).

**Comment by Yevgnen** [comment]
Thanks for you information! I didn't know about that. I should read more in the wiki before starting to hack into it :)

**Comment by Fuco1** [comment]
@Yevgnen we can add those two as aliases, if you would like that open a new issue (or a pull request :))
